### PR TITLE
feat(codepipeline): support cross-environment deployments for all actions

### LINKED ISSUE #3389: Update & test all AWS-owned CodePipeline Actions to pass their backing resource
Actions that have the default `'AWS'` owner (other than the CloudFormation Actions, which work differently) have to pass the resource they are backed by up to the CodePipeline construct in order for the cross-region and cross-account functionality to work correctly.

Each of these needs a small code change (modifying the call to `super` to pass the `resource` property), and, more importantly, a manual test that verifies a cross-environment Pipeline including them works like expected. This test part is very important, as it's quite common to run into many subtle issues that are difficult to predict without actually using the Actions in a cross-environment manner (like the need to create inner resources with `PhysicalName.GENERATE_IF_NEEDED`, or the fact that source Actions require cross-account CloudWatch Events).

* [x] CodeCommit source Action
* [x] S3 source Action
* [x] ECR source Action
* [x] Lambda invoke Action
* [ ] CodeDeploy Server deploy Action
* [x] S3 deploy Action
* [ ] ECS deploy Action

(CodeBuild had been already updated & tested as the first, "beta" cross-account Action)

### PR DESCRIPTION
**Pull request by skinny85**
Previously, we only supported cross-environment deployments for CodeBuild
and CloudFormation CodePipeline actions.
This change adds this capability to all remaining AWS-owned actions.

Fixes #3389

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by skinny85** [comment]
There is an interesting aspect to this PR. The CodePipeline API has a very short limit on the ArtifactStore.EncryptionKey.Id property (just 100 characters). That is problematic when using Alias ARNs in there, as an Alias can have just its name be up to 256 characters long. I've opened an issue about it to the CodePipeline team, and worked around this problem in this PR ([here](https://github.com/aws/aws-cdk/pull/4276/files#diff-50ca3157547e730ae9c6cb99687ee0d3R20)). Depending on what we think is better, we can wait until the issue is fixed, and remove that workaround, or merge this as-is, and later change the generated Alias name.

File: packages/@aws-cdk/aws-events/lib/rule.ts [hunk 1]
```diff
@@ -229,12 +229,14 @@ export class Rule extends Resource implements IRule {
...
-        new Rule(targetStack, `${this.node.uniqueId}-${id}`, {
+        const rule = new Rule(targetStack, `${this.node.uniqueId}-${id}`, {
           targets: [target],
-          eventPattern: this.eventPattern,
           schedule: this.scheduleExpression ? Schedule.expression(this.scheduleExpression) : undefined,
           description: this.description,
         });
+        // eventPattern is mutable through addEventPattern(), so we need to lazy evaluate it
+        // but only Tokens can be lazy in the framework, so do this magical hack instead
+        (rule as any).eventPattern = this.eventPattern;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events/lib/rule.ts]: Not entirely sure why this is the way it is. Is it to prevent mutation? Couldn't we leave users to shoot themselves in the foot?

File: packages/@aws-cdk/aws-kms/lib/alias.ts [hunk 2]
```diff
@@ -160,9 +173,19 @@ export class Alias extends AliasBase {
...
+    const maxLength = this._generatedNameMaxLength - REQUIRED_ALIAS_PREFIX.length;
+    // take the last characters, as they include the hash,
+    // and so have a higher chance of not colliding
+    return REQUIRED_ALIAS_PREFIX + lastNCharacters(super.generatePhysicalName(), maxLength);
   }
 }
+
+function lastNCharacters(str: string, n: number) {
+  const startIndex = Math.max(str.length - n, 0);
+  return str.substring(startIndex, str.length);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-kms/lib/alias.ts]: 2nd parameter is optional :)

File: packages/@aws-cdk/aws-codepipeline-actions/lib/codecommit/source-action.ts [hunk 3]
```diff
@@ -68,6 +68,7 @@ export class CodeCommitSourceAction extends Action {
 
     super({
       ...props,
+      resource: props.repository,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-codepipeline-actions/lib/codecommit/source-action.ts]: Is this what makes cross-stack magic work?

File: packages/@aws-cdk/aws-events/lib/rule.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-events/lib/rule.ts]: This is to make it that both rules (the original, and the new one created in the target's account) have exactly the same `eventPattern`. Because of mutability with `addEventPattern()`, just passing `this.eventPattern` in the constructor of the other rule does not provide that. Hence this brilliant solution/horrible hack (whichever you prefer) to make sure the 2 are identical.

File: packages/@aws-cdk/aws-kms/lib/alias.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-kms/lib/alias.ts]: Sure, will change.

File: packages/@aws-cdk/aws-codepipeline-actions/lib/codecommit/source-action.ts (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-actions/lib/codecommit/source-action.ts]: Yep.

**Comment by skinny85** [comment]
Rebased & and added a local cache of account to `Pipeline`.

File: packages/@aws-cdk/aws-kms/lib/alias.ts [hunk 4]
```diff
@@ -160,9 +173,19 @@ export class Alias extends AliasBase {
     if (props.removalPolicy) {
       resource.applyRemovalPolicy(props.removalPolicy);
     }
+
+    this._generatedNameMaxLength = props.generatedNameMaxLength || 256;
   }
 
   protected generatePhysicalName(): string {
```
Comment (eladb) [inline packages/@aws-cdk/aws-kms/lib/alias.ts]: Can you subclass `Alias` and implement this locally in the CodePipeline codebase so this hack doesn't leak into the public API?

Comment (skinny85) [reply packages/@aws-cdk/aws-kms/lib/alias.ts]: Done.

**Comment by skinny85** [comment]
Moved the Alias hack to the CodePipeline module, according to Elad's comment, and rebased.

File: packages/@aws-cdk/aws-events/test/test.rule.ts [hunk 5]
```diff
@@ -607,11 +607,7 @@ export = {
           region: 'us-west-2',
         },
       });
-      const rule = new Rule(sourceStack, 'Rule', {
-        eventPattern: {
-          source: ['some-event'],
-        },
-      });
+      const rule = new Rule(sourceStack, 'Rule');
```
Comment (eladb) [inline packages/@aws-cdk/aws-events/test/test.rule.ts]: Is this a regression in behavior? Why was this needed to change? I’d expect this test to remain in tact. If there’s another case you are trying to cover, please add an additional test.

Comment (skinny85) [reply packages/@aws-cdk/aws-events/test/test.rule.ts]: The test was not written in an exhaustive enough way before - in this new form, it fails without the production changes, even though the assertions stayed the same.

Comment (eladb) [reply packages/@aws-cdk/aws-events/test/test.rule.ts]: Not sure what that means... as far as I understand the previous code should work and therefore the old test is relevant. If it now fails, then something is wrong. If you believe we need additional coverage, please add an additional test.

Comment (skinny85) [reply packages/@aws-cdk/aws-events/test/test.rule.ts]: Added a new test.

File: packages/@aws-cdk/aws-events/lib/rule.ts [hunk 6]
```diff
@@ -229,12 +229,14 @@ export class Rule extends Resource implements IRule {
...
         // The actual rule lives in the target stack.
         // Other than the account, it's identical to this one
-        new Rule(targetStack, `${this.node.uniqueId}-${id}`, {
+        const rule = new Rule(targetStack, `${this.node.uniqueId}-${id}`, {
           targets: [target],
-          eventPattern: this.eventPattern,
           schedule: this.scheduleExpression ? Schedule.expression(this.scheduleExpression) : undefined,
           description: this.description,
         });
+        // eventPattern is mutable through addEventPattern(), so we need to lazy evaluate it
```
Comment (eladb) [inline packages/@aws-cdk/aws-events/lib/rule.ts]: Why not use Lazy.asAny() and avoid hackery?

Comment (eladb) [reply packages/@aws-cdk/aws-events/lib/rule.ts]: Alternatively expose a proper @internal API for this. Using “as any” should be a very very rare thing if at all.

Comment (skinny85) [reply packages/@aws-cdk/aws-events/lib/rule.ts]: Changed to a subclass and made `@internal`.

**Comment by skinny85** [comment]
Done like we talked about. I also needed to override the `validate` method in the inner class, but that makes sense when you think about it (otherwise, all validation errors for that rule would be duplicated).
